### ISSUE TITLE
Unchecked Exception is not thrown for http status 404

### ISSUE THREAD

**Issue by pravinrajr9** [ContainX/openstack4j#1180]
The execute method of BaseOpenStackService class does not throw exception for http response status 404.
Instead an error message is logged. It would be better if it had thrown exception for illegal access. So that the one who is using this client would understand the cause more faster.  Currently it is returning null as response for any http response with 404 status.
Is it left for logging for any use case or it has to be fixed?

I would also be glad to fix and contribute for openstack4j

**Comment by pravinrajr9** [comment]
@gondor , @auhlig I did not get purpose of not throwing an exception for response with http 404 status code. I would be glad to fix this, if it is expected to throw an exception

**Comment by vinodborole** [comment]
Throwing exception or checking specific status code is not required with the core library as it will not be generic, the Response Object returned is sufficient for the caller to make use of it they way it wants.

**Comment by pravinrajr9** [comment]
@vinodborole Wondering why there is no exception thrown for only response having http error code 404.
While other error codes less then 500 and not equal to 404, an exception is thrown. The response object returned is null and again we need to make an extra call to know the actual cause.

**Comment by vinodborole** [comment]
its handled here - https://github.com/ContainX/openstack4j/blob/master/core/src/main/java/org/openstack4j/core/transport/HttpExceptionHandler.java

Please go through the implementation for handling exceptions.

**Comment by pravinrajr9** [comment]
@vinodborole I agree but in the HttpEntityHandler  line no 43 returns true so it never reach the line                ` throw mapException(response.getStatusMessage(), response.getStatus());`  where it calls mapExection in HttpExceptionHandler.java 
 
So it is returning null instead of throwing an exception.

**Comment by pravinrajr9** [comment]
@vinodborole  In method handle404 below

```
    private static <T> Handle<T> handle404(Handle<T> handle) {
        if (handle.getResponse().getStatus() == 404) {

            if (ListType.class.isAssignableFrom(handle.getReturnType())) {
                try {
                    return handle.complete(handle.getReturnType().newInstance());
                } catch (InstantiationException e) {
                    LOG.error(e.getMessage(), e);
                } catch (IllegalAccessException e) {
                    LOG.error(e.getMessage(), e);
                }
            }

            if (handle.getReturnType() != ActionResponse.class) {
                return handle.complete(null);
            }
        }

        return handle.continueHandling();
    }
```

I think, not sure much, there might be problem, in the if condition and  comparing for ActionResponse class != handle.getReturnType(). 
Below one may give expected result.

 ```
private static <T> Handle<T> handle404(Handle<T> handle) {
        if (handle.getResponse().getStatus() == 404) {

            if (ListType.class.isAssignableFrom(handle.getReturnType())) {
                try {
                    return handle.complete(handle.getReturnType().newInstance());
                } catch (InstantiationException e) {
                    LOG.error(e.getMessage(), e);
                } catch (IllegalAccessException e) {
                    LOG.error(e.getMessage(), e);
                }
            }

            if (handle.getReturnType() == ActionResponse.class) {
                return handle.complete(null);
            }
        }

        return handle.continueHandling();
    }
```

Please have a look and let us know

**Comment by dgerenrot** [comment]
That way, you would not return action response on 404, but you would return
something in every other case. I think what they are doing does make sense.
It may also make sense to get rid of that return(null) altogether.

**Comment by pravinrajr9** [comment]
@dgerenrot In our case the handle.complete call returns true since handle.getReturnType is not of type ActionResponse. So the exception it wont reach the mapException call in HttpEntityHandler.java
 
I think it should be something like below:
` ActionResponse ar = ResponseToActionResponse.INSTANCE.apply(handle.getResponse());
                if (handle.getReturnType() == ActionResponse.class) {
                    return handle.complete((T) ar);
                }`

**Comment by pravinrajr9** [comment]
I tried to debug further and there are some test cases which expects null instead of getting unchecked exception for example the code in org.openstack4j.api.compute.ServerTests the method getServerConsoleOutputNonExistingServer() is expecting null object instead of getting an run time exception. 

And also in the http://openstack4j.com/learn

![image](https://user-images.githubusercontent.com/30862177/39567711-81d03abc-4edd-11e8-9a83-8b148b7eae6e.png)

It says ClientResponseException is thrown for error between 400 and 409 excluding 404.
I think it would better to mention in below line to include both checked and unchecked exceptions instead of only checked exception 

![image](https://user-images.githubusercontent.com/30862177/39567838-df78bacc-4edd-11e8-9c64-286a24e291a9.png)
@gondor @frsyuki @vinodborole Please put on your views
